Add script to automatically update changelog

charlesbvll opened the pull request:
<!--
Thank you for opening a pull request (PR)!

Contribution guidelines: https://github.com/adap/flower/blob/main/CONTRIBUTING.md
-->

## Issue

### Description

Currently we have to manually update the changelog with all new changes since the previous release. This is very tedious.

### Related issues/PRs

N/A

## Proposal

### Explanation

This new script allows us to add changelog entries based on all new PRs' descriptions.
By default, it will search for a `Changelog entry` section in the description of the PR, if it is empty, it will just use the title of the PR for the changelog entry, without any description. If it is non existent, it will categorize the PR as "General improvement" and add it to the changelog accordingly. If it exists, it will use it to add a description to the change.

Note that a system of tokens is also in place:
- `<general>` is for classifying a PR as a general improvement.
- `<skip>` is to not add the PR to the changelog
- `<baselines>` is to add a general baselines change to the PR
- `<examples>` is to add a general examples change to the PR
- `<sdk>` is to add a general sdk change to the PR
- `<simulations>` is to add a general simulations change to the PR

Once those tokens are detected, if the corresponding entry (like `- **General improvements**`) doesn't already exist, it gets automatically created, and otherwise, PRs' URLs are added at the end of the list inside the parenthesis.

### Checklist

- [x] Implement proposed change
- [x] Update changelog entry below
- [x] Make CI checks pass
- [x] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### Changelog entry

<skip>

### Any other comments?

In the new PR template, <general> is the default flag.

Robert-Steiner commented on src/py/flwr_tool/update_changelog.py:
```diff
@@ -0,0 +1,225 @@
+    content, category_title, pr_info, unreleased_index, next_header_index
+):
+    """Update a specific section in the changelog content."""
+    section_index = content.find(category_title, unreleased_index, next_header_index)
+    if section_index != -1:
```
just want to mention a small tick i have discovered recently. the two lines can be combined in one by using the walrus operator `:=` that has been introduced in python 3.8.

(i haven't tested this line)
```suggestion
    if (section_index := content.find(category_title, unreleased_index, next_header_index)) != -1:
```
but i have no strong preference using it here so we can also leave the code as is.

charlesbvll replied:
Oh nice! I'll test it out, thanks!

Robert-Steiner reviewed (commented):
Great work! I really like the script 👍

charlesbvll commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -41,6 +41,10 @@ Example: The variable `rnd` was renamed to `server_round` to improve readability
 
+## Changelog entry
+
+<general>
```
That's a good point, I will add a comment explaining everything above the section

tanertopal commented:
@charlesbvll This is cool! I believe it would be helpful to write a small guide in the docs "CONTRIBUTOR HOW-TO GUIDES" section on how this works, with a few examples to send to new contributors to explain it.

@danieljanes @Robert-Steiner thoughts?

charlesbvll commented:
> @charlesbvll This is cool! I believe it would be helpful to write a small guide in the docs "CONTRIBUTOR HOW-TO GUIDES" section on how this works, with a few examples to send to new contributors to explain it.
> 
> @danieljanes @Robert-Steiner thoughts?

I just added a quick mention at the moment, where we talk about the description of the PR, but maybe I can add a new section at the end of the doc, like an appendix, giving more details and some examples.

charlesbvll commented:
> > @charlesbvll This is cool! I believe it would be helpful to write a small guide in the docs "CONTRIBUTOR HOW-TO GUIDES" section on how this works, with a few examples to send to new contributors to explain it.
> > @danieljanes @Robert-Steiner thoughts?
> 
> I just added a quick mention at the moment, where we talk about the description of the PR, but maybe I can add a new section at the end of the doc, like an appendix, giving more details and some examples.

@tanertopal I added an appendix, I think the PR is ready for review

Robert-Steiner reviewed (commented):
From my side it looks good!
